File: GCDLib/Core/Visualization/ElevationChangeDataPoint.vb
Namespace Core.Visualization

    Public Class ElevationChangeDataPoint

        Private m_fElevation As Double
        Private m_fAreaChange As Double
        Private m_fVolumeChange As Double

        Public ReadOnly Property Elevation As Double
            Get
                Return m_fElevation
            End Get
        End Property

        Public Sub New(fElevation As Double, fArea As Double, fVolume As Double)
            m_fElevation = fElevation
            m_fAreaChange = fArea
            m_fVolumeChange = fVolume
        End Sub

        Public Function AreaChange(dataUnits As UnitsNet.Units.AreaUnit, displayUnits As UnitsNet.Units.AreaUnit) As Double
            Return UnitsNet.Area.From(m_fAreaChange, dataUnits).As(displayUnits)
        End Function

        Public Function VolumeChange(dataUnits As UnitsNet.Units.VolumeUnit, displayUnits As UnitsNet.Units.VolumeUnit) As Double
            Return UnitsNet.Volume.From(m_fVolumeChange, dataUnits).As(displayUnits)
        End Function

        Public Function GetElevation(dataUnits As UnitsNet.Units.LengthUnit, displayUnits As UnitsNet.Units.LengthUnit) As Double
            Return UnitsNet.Length.From(m_fElevation, dataUnits).As(displayUnits)
        End Function

        Public Function AreaDeposition(dataUnits As UnitsNet.Units.AreaUnit, displayUnits As UnitsNet.Units.AreaUnit) As Double

            If m_fElevation > 0 Then
                Return UnitsNet.Area.From(m_fAreaChange, dataUnits).As(displayUnits)
            Else
                Return 0
            End If

        End Function

        Public Function AreaErosion(dataUnits As UnitsNet.Units.AreaUnit, displayUnits As UnitsNet.Units.AreaUnit) As Double
            If m_fElevation < 0 Then
                Return UnitsNet.Area.From(m_fAreaChange, dataUnits).As(displayUnits)
            Else
                Return 0
            End If
        End Function

        Public Function VolumeErosion(dataUnits As UnitsNet.Units.VolumeUnit, displayUnits As UnitsNet.Units.VolumeUnit) As Double
            If m_fElevation < 0 Then
                Return UnitsNet.Volume.From(m_fVolumeChange, dataUnits).As(displayUnits)
            Else
                Return 0
            End If
        End Function

        Public Function VolumeDeposition(dataUnits As UnitsNet.Units.VolumeUnit, displayUnits As UnitsNet.Units.VolumeUnit) As Double
            If m_fElevation > 0 Then
                Return UnitsNet.Volume.From(m_fVolumeChange, dataUnits).As(displayUnits)
            Else
                Return 0
            End If
        End Function

    End Class

End Namespace

File: GCDCore/Visualization/ElevationChangeDataPoint.cs
using System;
using System.Collections.Generic;

namespace GCDCore.Visualization
{
    public class ElevationChangeDataPoint
    {
        private double m_fElevation;
        private double m_fAreaChange;

        private double m_fVolumeChange;
        public double Elevation
        {
            get { return m_fElevation; }
        }

        public ElevationChangeDataPoint(double fElevation, double fArea, double fVolume)
        {
            m_fElevation = fElevation;
            m_fAreaChange = fArea;
            m_fVolumeChange = fVolume;
        }

        public double AreaChange(UnitsNet.Units.AreaUnit dataUnits, UnitsNet.Units.AreaUnit displayUnits)
        {
            return UnitsNet.Area.From(m_fAreaChange, dataUnits).As(displayUnits);
        }

        public double VolumeChange(UnitsNet.Units.VolumeUnit dataUnits, UnitsNet.Units.VolumeUnit displayUnits)
        {
            return UnitsNet.Volume.From(m_fVolumeChange, dataUnits).As(displayUnits);
        }

        public double GetElevation(UnitsNet.Units.LengthUnit dataUnits, UnitsNet.Units.LengthUnit displayUnits)
        {
            return UnitsNet.Length.From(m_fElevation, dataUnits).As(displayUnits);
        }

        public double AreaDeposition(UnitsNet.Units.AreaUnit dataUnits, UnitsNet.Units.AreaUnit displayUnits)
        {
            if (m_fElevation > 0)
            {
                return UnitsNet.Area.From(m_fAreaChange, dataUnits).As(displayUnits);
            }
            else
            {
                return 0;
            }
        }

        public double AreaErosion(UnitsNet.Units.AreaUnit dataUnits, UnitsNet.Units.AreaUnit displayUnits)
        {
            if (m_fElevation < 0)
            {
                return UnitsNet.Area.From(m_fAreaChange, dataUnits).As(displayUnits);
            }
            else
            {
                return 0;
            }
        }

        public double VolumeErosion(UnitsNet.Units.VolumeUnit dataUnits, UnitsNet.Units.VolumeUnit displayUnits)
        {
            if (m_fElevation < 0)
            {
                return UnitsNet.Volume.From(m_fVolumeChange, dataUnits).As(displayUnits);
            }
            else
            {
                return 0;
            }
        }

        public double VolumeDeposition(UnitsNet.Units.VolumeUnit dataUnits, UnitsNet.Units.VolumeUnit displayUnits)
        {
            if (m_fElevation > 0)
            {
                return UnitsNet.Volume.From(m_fVolumeChange, dataUnits).As(displayUnits);
            }
            else
            {
                return 0;
            }
        }
    }
}